cps/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -767,10 +786,31 @@ synchronized FlowHead getFirstHead() {
         return heads.firstEntry().getValue();
     }
 
-    void notifyListeners(FlowNode node) {
-        if (listeners != null) {
-            for (GraphListener listener : listeners) {
-                listener.onNewHead(node);
+    void notifyListeners(List<FlowNode> nodes, boolean synchronous) {
+        List<GraphListener> _listeners = synchronous ? synchronousListeners : listeners;
+        if (_listeners != null) {
+            Saveable s = Saveable.NOOP;
+            try {
+                Queue.Executable exec = owner.getExecutable();
+                if (exec instanceof Saveable) {
+                    s = (Saveable) exec;
+                }
+            } catch (IOException x) {
+                LOGGER.log(Level.WARNING, null, x);
+            }
+            BulkChange bc = new BulkChange(s);
+            try {
+                for (FlowNode node : nodes) {
+                    for (GraphListener listener : _listeners) {
+                        listener.onNewHead(node);
+                    }
+                }
+            } finally {
+                try {
+                    bc.commit();
+                } catch (IOException x) {
+                    LOGGER.log(Level.WARNING, null, x);
+                }
```

Comment:
They are not being treated differently here. One or the other is selected according to the method parameter.